src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/PageSaveTempDataPropertyFilter.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public class PageSaveTempDataPropertyFilter : SaveTempDataPropertyFilterBase
+    {
+        public PageSaveTempDataPropertyFilter(ITempDataDictionaryFactory factory)
+            : base(factory)
+        {
+        }
+
+        public PageSaveTempDataPropertyFilterFactory FilterFactory { get; set; }
+
+        /// <summary>
+        /// Creates and applies the set of <see cref="TempDataProperty"/>'s for the given <paramref name="modelType"/>.
+        /// </summary>
+        /// <param name="modelType">The <see cref="Type"/> to create <see cref="TempDataProperty"/>'s for.</param>
+        public void SetTempDataProperties(Type modelType)
+        {
+            if (FilterFactory == null)
+            {
+                throw new ArgumentNullException(nameof(FilterFactory));
+            }
+
+            TempDataProperties = FilterFactory.GetTempDataProperties(modelType);
+        }
```

Comment:
One thing that you could do here is you could make `Subject` virtual, and you could do this work when the property is set.